// parse-handler/index.js
if (Parse == null) var Parse = require('parse').Parse;

var BUSINESS_TABLE = 'Business';
var REWARDS_TABLE = 'Reward';
var INTERVAL_TABLE = 'NEW_Interval';
var INTERVAL_RECORD_TABLE = 'NEW_IntervalRecord';

/*
args list format:
  { 'arg1': arg1Obj, 'arg2': arg2Obj }

success callback format:
  function (object) {
    ...
  }

failure callback format:
  function (object, error) {
    ...
    console.log(error.message);
  }

*/

// arguments must include parse user object
var retrieveBusiness = function (successCb, failureCb, args) {
  var user = args['user'];

  var BusinessObj = Parse.Object.extend(BUSINESS_TABLE);
  var businessQuery = new Parse.Query(BusinessObj);

  console.log('retrieving business');

  businessQuery.equalTo('owner', user);
  businessQuery.first({
    success: successCb,
    error: failureCb
  })
}

// arguments must include business object
var retrieveRewards = function (successCb, failureCb, args) {
  var business = args['business'];

  var RewardObj = Parse.Object.extend(REWARDS_TABLE);
  var rewardQuery = new Parse.Query(RewardObj);

  rewardQuery.equalTo('business', business);

  rewardQuery.find({
    success: successCb,
    error: failureCb
  });
}


// arguments must include interval object for a user
var retrieveIntervalJSON = function (successCb, failureCb, args) {
  var user = args['user'];

  // if business is successfully queried, then query interval table
  var retBusSuccess = function (business) {
    var IntervalObj = Parse.Object.extend(INTERVAL_TABLE);
    var intervalQuery = new Parse.Query(IntervalObj);

    console.log(JSON.stringify(business));

    intervalQuery.find({
      success: successCb,
      error: failureCb
    });
  }

  var retBusFailure = function (business, error) {
    console.log('ERROR:'+error);
  }

  retrieveBusiness(retBusSuccess, retBusFailure, {'user': user});
}

// arguments must include the current parse user
// returns a list where each element is a list of intervals
var retrieveIntervalRecordsJSON = function (successCb, failureCb, args) {
    var user = args['user'];

    var intervalSuccess = function (intervals) {
      
      var intervalQueryList = [];

      intervals.forEach(function (interval) {
        var IntervalRecordObj = Parse.Object.extend(INTERVAL_RECORD_TABLE);
        var intervalRecordQuery = new Parse.Query(IntervalRecordObj);

        intervalRecordQuery.equalTo('interval', interval);

        var q = intervalRecordQuery.find();

        intervalQueryList.push(q);
      });
      
      Parse.Promise.when(intervalQueryList).then(function() {
        successCb(arguments);
        //console.log(JSON.stringify(arguments));
      });
    }
    
    retrieveIntervalJSON(intervalSuccess, failureCb, {'user': user});
}

exports.retrieveBusiness = retrieveBusiness;
exports.retrieveRewards = retrieveRewards;
exports.retrieveIntervalJSON = retrieveIntervalJSON;
exports.retrieveIntervalRecordsJSON = retrieveIntervalRecordsJSON;
